adapters/adagio/adagio.go

```diff
@@ -0,0 +1,139 @@
+	json, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	if request.Test == 0 {
+		// Gzip the body
+		// Note: Gzipping could be handled natively later: https://github.com/prebid/prebid-server/issues/1812
+		var bodyBuf bytes.Buffer
+		gz := gzip.NewWriter(&bodyBuf)
+		if _, err = gz.Write(json); err == nil {
+			if err = gz.Close(); err == nil {
+				json = bodyBuf.Bytes()
+				headers.Add("Content-Encoding", "gzip")
+				// /!\ Go already sets the `Accept-Encoding: gzip` header. Never add it manually, or Go won't decompress the response.
+				//headers.Add("Accept-Encoding", "gzip")
+			}
+		}
+	}
+
+	requestToBidder := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    json,
+		Headers: headers,
+	}
+
+	return []*adapters.RequestData{requestToBidder}, nil
+}
+
+const unexpectedStatusCodeFormat = "Unexpected status code: %d. Run with request.debug = 1 for more info"
+
+// MakeBids unpacks the server's response into Bids.
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, _ *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	switch response.StatusCode {
+	case http.StatusOK:
+		break
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusServiceUnavailable:
+		fallthrough
+	case http.StatusBadRequest:
+		fallthrough
+	case http.StatusUnauthorized:
+		fallthrough
+	case http.StatusForbidden:
+		err := &errortypes.BadInput{
+			Message: fmt.Sprintf(unexpectedStatusCodeFormat, response.StatusCode),
+		}
+		return nil, []error{err}
+	default:
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf(unexpectedStatusCodeFormat, response.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var openRTBBidderResponse openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &openRTBBidderResponse); err != nil {
+		return nil, []error{err}
+	}
+
+	bidsCapacity := len(internalRequest.Imp)
+	errs := make([]error, 0, bidsCapacity)
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(bidsCapacity)
+	var typedBid *adapters.TypedBid
+	for _, seatBid := range openRTBBidderResponse.SeatBid {
+		for _, bid := range seatBid.Bid {
+			activeBid := bid
+
+			activeExt := &extBid{}
+			if err := json.Unmarshal(activeBid.Ext, activeExt); err != nil {
+				errs = append(errs, err)
+			}
+
+			var bidType openrtb_ext.BidType
+			if activeExt.Prebid != nil && activeExt.Prebid.Type != "" {
+				bidType = activeExt.Prebid.Type
+			} else {
+				err := &errortypes.BadInput{
```

Comment:
The error code in this case should be `BadServerResponse` as the bidder response doesn't include the bidType. `BadInput` indicates that the request was malformed or was missing some required info.